File: RailgunVB/Commands/Music/MusicAuto.vb
Imports Discord
Imports Discord.Commands
Imports Microsoft.Extensions.DependencyInjection
Imports RailgunVB.Core
Imports RailgunVB.Core.Preconditions
Imports TreeDiagram
Imports TreeDiagram.Models.Server

Namespace Commands.Music
    
    Partial Public Class Music
        
        <Group("auto"), UserPerms(GuildPermission.ManageGuild)>
        Public Class MusicAuto
            Inherits SystemBase
            
            <Command("join")>
            Public Async Function JoinAsync() As Task
                Dim data As ServerMusic = Await Context.Database.ServerMusics.GetOrCreateAsync(Context.Guild.Id)
                Dim vc As IVoiceChannel = (CType(Context.User, IGuildUser)).VoiceChannel
                    
                If vc Is Nothing AndAlso data.AutoVoiceChannel = 0
                    Await ReplyAsync("Music Auto-Join is currently disabled. Please join a voice channel and run this command again to enable it.")
                    Return
                ElseIf vc Is Nothing AndAlso data.AutoVoiceChannel <> 0
                    data.AutoVoiceChannel = 0
                    data.AutoTextChannel = 0
                        
                    await ReplyAsync("Music Auto-Join has been disabled.")
                    Return
                End If
                    
                data.AutoVoiceChannel = vc.Id
                data.AutoTextChannel = Context.Channel.Id
                    
                await ReplyAsync($"{If(data.AutoVoiceChannel = 0, "Music Auto-Join is now enabled!", "") _ 
                    } Will automatically join {Format.Bold(vc.Name)} and use {Format.Bold(
                        "#" + Context.Channel.Name)} to post status messages.")
            End Function
            
            <Command("skip")>
            Public Async Function SkipAsync() As Task
                Dim data As ServerMusic = Await Context.Database.ServerMusics.GetOrCreateAsync(Context.Guild.Id)
                
                data.AutoSkip = Not (data.AutoSkip)
                
                await ReplyAsync($"Music Auto-Skip is now {Format.Bold(If(data.AutoSkip, "Enabled", "Disabled"))}.")
            End Function
            
            <Command("download")>
            Public Async Function DownloadAsync() As Task
                Dim data As ServerMusic = Await Context.Database.ServerMusics.GetOrCreateAsync(Context.Guild.Id)
                
                data.AutoDownload = Not (data.AutoDownload)
                
                Await ReplyAsync($"Music Auto-Download is now {Format.Bold(If(data.AutoDownload, "Enabled", 
                                                                              "Disabled"))}.")
            End Function
            
        End Class
        
    End Class
    
End NameSpace

File: Railgun/Commands/Music/MusicAuto.cs
using System.Threading.Tasks;
using AudioChord;
using Discord;
using Finite.Commands;
using Railgun.Core;
using Railgun.Core.Attributes;
using Railgun.Music;
using TreeDiagram;

namespace Railgun.Commands.Music
{
    public partial class Music
	{
		[Alias("auto"), UserPerms(GuildPermission.ManageGuild)]
		public partial class MusicAuto : SystemBase
		{
			private readonly PlayerController _playerController;
			private readonly MusicService _music;

			public MusicAuto(PlayerController playerController, MusicService music)
			{
				_playerController = playerController;
				_music = music;
			}

			[Command("play")]
			public async Task PlayAsync(string input)
			{
				var profile = Context.Database.ServerProfiles.GetOrCreateData(Context.Guild.Id);
            	var data = profile.Music;
				var playlist = await SystemUtilities.GetPlaylistAsync(_music, data);

				SongId songId;
				var videoId = YoutubeExplode.Videos.VideoId.TryParse(input);

				if (input.Contains("http") && videoId != null)
					songId = SongId.Parse("YOUTUBE#" + videoId.Value);
				else
					songId = SongId.Parse(input);

				if (!playlist.Songs.Contains(songId))
				{
					await ReplyAsync("Unable to find song in playlist! Please add it to playlist.");
					return;
				}
				if (data.AutoPlaySong == songId.ToString())
				{
					await ReplyAsync("This song is already set to play on auto-join.");
					return;
				}

				data.AutoPlaySong = songId.ToString();
				await ReplyAsync("Will now play specified song on auto-join!");
			}

			[Command("play")]
			public Task PlayAsync()
			{
				var profile = Context.Database.ServerProfiles.GetOrCreateData(Context.Guild.Id);
            	var data = profile.Music;

				data.AutoPlaySong = string.Empty;
				return ReplyAsync("Will no longer play specific song on auto-join.");
			}

			[Command("skip")]
			public Task SkipAsync()
			{
				var profile = Context.Database.ServerProfiles.GetOrCreateData(Context.Guild.Id);
            	var data = profile.Music;

				data.AutoSkip = !data.AutoSkip;
				return ReplyAsync($"Music Auto-Skip is now {Format.Bold(data.AutoSkip ? "Enabled" : "Disabled")}.");
			}
		}
	}
}